### ISSUE TITLE
error handling with no checked exceptions

### ISSUE THREAD

**Issue by migueltorcha** [mobilejazz/harmony-kotlin#66]
My proposal is basically to stop using exceptions in benefit of the functional approach monads. In kotlin, like in many other modern languages, we do not have checked exceptions, so it is not mandatory to catch any exception, and even worst, you probably won’t be aware that some exceptions could be thrown. This is making us, sometimes, to catch a generic Exception, so anything (expected or not) could be catch and hidden there.
With my proposed approach, exceptions would be used just for “exceptional” situations, and we should avoid them to handle domain errors.
What does relevant people think about this? Well, I’ve found some interesting articles with many arguments that could clarify this issue:

**Why should we avoid checked exceptions?**
- Kotlin documentation:
https://kotlinlang.org/docs/exceptions.html#checked-exceptions (take also a look to inner links)
- Google Testing Blog:
https://testing.googleblog.com/2009/09/checked-exceptions-i-love-you-but-you.html
- Robert C. Martin (Clean code Author):
https://pastebin.com/Rfzx3P2L (I’ve extracted this from his book, where he dedicates one whole chapter to Error handling)

**Any alternative to checked exceptions?**
- Roman Elizarov (Project Lead for the Kotlin Programming Language):
https://elizarov.medium.com/kotlin-and-exceptions-8062f589d07
- Arrow (funcional library based in kotlin)
https://arrow-kt.io/docs/0.11/apidocs/arrow-core-data/arrow.core/-either/#using-either-instead-of-exceptions
- Adam Bennett (Google developer Expert)
https://adambennett.dev/2020/05/the-result-monad/
- Martin Fowler (software development books writer and speaker):
https://martinfowler.com/articles/replaceThrowWithNotification.html

**Comment by Alex-DA** [comment]
This is out of my scope but it seems interesting, I stay here waiting for the discussion.

**Comment by migueltorcha** [comment]
One more interesting reading about this: https://proandroiddev.com/considering-all-unhappy-paths-in-a-type-safe-way-in-modern-android-bc41e9aaa55

**Comment by AresProductions** [comment]
Personally, I've used the "functional" approach of error handling. Specifically, I've used it when doing reactive programming on RxJava, Kotlin Flow, and Combine. Actually, Swift 5.0 does give native support to such a pattern via the Future, Publisher, and Result classes.

The way I've used it is to throw errors on the very basic data sources (for example 404 response, database failing a delete/insert request) which then is mapped to a Result type. 

So no repositories or interactors would ever throw or use try-catch statements. They would only use and return Result types that could be mapped accordingly.

The way I'm using the Result type may not be optimal, but the general idea is to always handle expected error cases on the Domain, Repo, and Presentation layer.
